python/ray/data/_internal/execution/bulk_executor.py

```diff
@@ -0,0 +1,99 @@
+from ray.data._internal.stats import DatasetStats
+
+logger = logging.getLogger(__name__)
+
+
+class BulkExecutor(Executor):
+    def __init__(self, options: ExecutionOptions):
+        super().__init__(options)
+        self._stats = DatasetStats(stages={}, parent=None)
+        self._executed = False
+
+    def execute(
+        self, dag: PhysicalOperator, initial_stats: Optional[DatasetStats] = None
+    ) -> Iterator[RefBundle]:
+        """Synchronously executes the DAG via bottom-up recursive traversal."""
+
+        assert not self._executed, "Can only call execute once."
+        self._executed = True
+        if not isinstance(dag, InputDataBuffer):
+            logger.info("Executing DAG %s", dag)
+
+        if initial_stats:
+            self._stats = initial_stats
+
+        saved_outputs: Dict[PhysicalOperator, List[RefBundle]] = {}
+
+        def execute_recursive(op: PhysicalOperator) -> List[RefBundle]:
+            # Avoid duplicate executions.
+            if op in saved_outputs:
+                return saved_outputs[op]
+
+            # Compute dependencies.
+            inputs = [execute_recursive(dep) for dep in op.input_dependencies]
+
+            # Fully execute this operator.
+            logger.debug("Executing op %s", op.name)
+            builder = self._stats.child_builder(op.name)
+            try:
+                for i, ref_bundles in enumerate(inputs):
+                    for r in ref_bundles:
+                        op.add_input(r, input_index=i)
+                    op.inputs_done(i)
+                output = _naive_run_until_complete(op)
+            finally:
+                op.shutdown()
+
+            # Cache and return output.
+            saved_outputs[op] = output
+            op_stats = op.get_stats()
+            op_metrics = op.get_metrics()
+            if op_stats:
+                self._stats = builder.build_multistage(op_stats)
+                self._stats.extra_metrics = op_metrics
+            return output
+
+        return execute_recursive(dag)
+
+    def get_stats(self) -> DatasetStats:
+        assert self._stats is not None, self._stats
+        return self._stats
+
+
+def _naive_run_until_complete(op: PhysicalOperator) -> List[RefBundle]:
+    """Run this operator until completion, assuming all inputs have been submitted.
+
+    Args:
+        op: The operator to run.
+
+    Returns:
+        The list of output ref bundles for the operator.
+    """
+    output = []
+    tasks = op.get_work_refs()
+    if tasks:
+        bar = ProgressBar(op.name, total=op.num_outputs_total())
+        while tasks:
+            done, _ = ray.wait(tasks, fetch_local=True, timeout=0.1)
+            for ready in done:
+                op.notify_work_completed(ready)
+            tasks = op.get_work_refs()
+            while op.has_next():
+                bar.update(1)
+                output.append(op.get_next())
+        bar.close()
+    while op.has_next():
+        output.append(op.get_next())
```

Comment:
Two questions:
1. Why is this needed? After the last of `op.get_work_refs()` has finished, all operator outputs should be done (in order, if needed), so the previous `while op.has_next()` loop should completely consume the operator outputs.
2. If this was needed, shouldn't we be updating the progress bar here?

If (1) is the case (i.e. this is not needed), this could be replaced with an assertion that this is in fact the case.
```suggestion
    assert not op.has_next()
```

Reply:
This is true for the Map operator, but other operators can have arbitrary behavior, such as no tasks at all.

Reply:
I get the no tasks case (and if that's the only case then maybe this should be in an `else:` block to make that clear), but if an operator does have tasks, I don't see how the last execution of the `while tasks` loop could consume all of the available operator outputs, and then have more operator outputs available right after exiting that loop, without any new operator event or state change.

>but other operators can have arbitrary behavior

Well, they can't have completely arbitrary behavior; the executor needs to have some contract around when it can expect the operator to have new outputs, and when it can expect the operator's output production to be finished. E.g. one sensible contract between the executor and the operator would be:
1. If an operator has 1 or more tasks, then it only produces more outputs after a task completes (i.e. after `op.notify_work_completed()` is called).
2. If an operator has 0 tasks, then it has a static set of outputs.

For (1), we try to consume the operator outputs whenever a task finishes; for (2), we do a single pass of operator output consumption. Whatever the contract, I'm just arguing for making it as explicit as possible in the executor code, so devs (like me) aren't left wondering "wait why would we need this last operator output consumption if the operator had tasks, shouldn't it have been exhausted at the end of the task loop?"

Reply:
Yes, good point. I added some comments in the docstring and here to clarify this completion behavior.

Reply:
It might be more clear to add a method to indicate this completion state of the operator. The added comment at has_next() and get_work_ref() seems a bit off-topic for them.